sdk/identity/Azure.Identity/src/ClientCertificateCredential.cs

```diff
@@ -135,5 +183,215 @@ public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext r
                 throw scope.FailAndWrap(e);
             }
         }
+
+        /// <summary>
+        /// IX509Certificate2Provider provides a way to control how the X509Certificate2 object is fetched.
+        /// </summary>
+        internal interface IX509Certificate2Provider
+        {
+            Task<X509Certificate2> GetCertificateAsync(bool async, CancellationToken cancellationToken);
```

Comment:
Should this return `ValueTask<X509Certificate2>` since it is only async in the case we need to read from a file, and then only on the first invocation?

Reply:
Great idea!  Updated.